design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,1111 @@
+
+<summary>Composition Tests</summary>
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// CompositionTest defines a test that runs a composition pipeline and
+// executes assertions on the resulting resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Namespaced,shortName=comptest,categories=meta
+type CompositionTest struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec CompositionTestSpec `json:"spec"`
+}
+
+// CompositionTestSpec defines the specification for the CompositionTest.
+//
+// +k8s:deepcopy-gen=true
+type CompositionTestSpec struct {
+	Tests []CompositionTestCase `json:"tests"`
+}
+
+// CompositionTestCase defines the specification of a single test case
+//
+// +k8s:deepcopy-gen=true
+type CompositionTestCase struct {
+	// Name of the Test Case, mandatory descriptive.
+	// Required.
+	Name string `json:"name"`
+
+	// ID is an optional unique identifier.
+	// Optional.
+	// +kubebuilder:validation:Optional
+	ID string `json:"id,omitempty"`
+
+	// Patches specifies patching configuration for the input XR.
+	// Optional.
+	// +kubebuilder:validation:Optional
+	Patches CompositionTestPatches `json:"patches,omitempty"`
+
+	// Inputs specifies the inputs paths or inline definitions.
+	// Required.
+	Inputs CompositionTestInputs `json:"inputs"`
+
+	// Assertions defines assertions to validate resources after test completion.
+	// Optional.
+	// +kubebuilder:validation:Optional
+	Assertions []runtime.RawExtension `json:"assertions,omitempty"`
+}
+
+// CompositionTestPatches defines the patches for a single test case
+//
+// +k8s:deepcopy-gen=true
+type CompositionTestPatches struct {
+	// XRD specifies the XRD.
+	XRD Resource `json:"xrd,omitempty"`
+
+	// AddFields specifies a map of fields:value that should be added to the input XR.
+	// Optional.
+	AddFields map[string]string `json:"addFields,omitempty"`
+
+	// RemoveFields specifies an array of fields that should be removed from the input XR.
+	// Optional.
+	RemoveFields []string `json:"removeFields,omitempty"`
+}
+
+type CompositionTestInputs struct {
+	// Timeout for the test in seconds
+	// Optional. Default is 30s.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=30
+	TimeoutSeconds *int `json:"timeoutSeconds"`
+
+	// Validate indicates whether to validate managed resources against schemas.
+	// Optional.
+	// +kubebuilder:validation:Optional
+	Validate *bool `json:"validate,omitempty"`
+
+	// XR specifies the composite resource.
+	XR Resource `json:"xr,omitempty"`
+
+	// Composition specifies the composition.
+	Composition Resource `json:"composition,omitempty"`
+
+	// Functions specifies the functions.
+	Functions []Resource `json:"functions,omitempty"`
```

Comment:
since we got rid of the duplicates, I think we should consider making those three mandatory, as they are the absolute minimum required params for `crossplane render`. Also, I suggest to add a CRDs field, or we could name it `Extensions` which is the term that `validate --help` is using

```suggestion
	// XR specifies the composite resource.
	XR Resource `json:"xr"`

	// Composition specifies the composition.
	Composition Resource `json:"composition"`

	// Functions specifies the functions.
	Functions []Resource `json:"functions"`
	
	// CRDs specifies the CRDs
	CRDs []Resource `json:"crds,omitempty"`
```

Reply:
Yes, that was my intent but forgot to remove the `omitempty`. Fixed.